docs/documentation/nodes.rst

```diff
@@ -0,0 +1,71 @@
+=====
+Nodes
+=====
+
+Nodes are the buildings blocks of the mathematical expressions represented by an individual.
+They are selected in the invididual's genome (see :ref:`pop_ind`).
+
+A node takes a certain number of inputs (specified by its `arity` which can also be 0) and transforms them into one output value. A priori, there are no limitations regarding the type of transformation, as long as it can be expressed in pure Python, SymPy_, PyTorch_, and NumPy_.
+
+----------------------------
+Implemented nodes in hal-cgp
+----------------------------
+
+hal-cgp comes with a set of implemented nodes:
+
+- The `InputNode` and `OutputNode` are special nodes that represent the input to and output from the computational graph.
+- A set of basic mathematical operations is implemented as operator nodes: addition (``Add``), subtraction (``Sub``), multiplication (``Mul``), division (``Div``) and power (``Pow``).
+- A `ConstantFloat` that represents a constant in mathematical expressions.
+- A `ParameterNode` that holds adjustable parameter (see :ref:`local_search`).
+
+------------
+Custom nodes
+------------
+
+In hal-cgp, it is straightforward to implement new, custom nodes.
+The custom nodes needs to be implemented as a subclass of the `OperatorNode` (:meth:`cgp.node.OperatorNode`) and define its ``arity`` as well as its transformation, defined by the ``_def_output`` member of the class. The ``_def_output`` string defines the transformation in pure Python and is used to define the transformation in SymPy_, PyTorch_, and Numpy_.
+
+
+As a simple example, we implement a custom node that doubles its input:
+
+  .. code-block:: python
+
+		  
+     import cgp
+     import math
```

Comment:
```suggestion
```

Reply:
Don't we need to import this package if we use it in the string?

Reply:
How does this work if we don't import it?

Reply:
man you're asking questions i can't answer ;) i think it's important that the packages are available when the function is compiled, i.e., in `cartesian_graph.py`. since the expression is not evaluated here i don't think you need the imports, could you give that a try please?

i wanted to look into this problem more deeply bc if we can provide a "compilation context" (context managers ftw?) for nodes, we might be able to define and use arbitrary functions in `_def_output`, sort of like:
```python
def myifelse(x_0, x_1, x_2):
    if x_0 < 0:
        return x_1
    else:
        return x_2

class MyIfElseNode(cgp.OperatorNode):
    _arity = 3
    _def_output = "myifelse(x_0, x_1, x_2)"
```

this would provide great flexibility:tm:

Reply:
ah, and for completeness, in this node you don't use `math` ;)

Reply:
Right, yes. Okay, the import is not required, indeed.
But yes, we have to find a solution for this, because relying on imports in `cartesian_graph.py` is not a good solution. Can you create an issue for that?

Reply:
done